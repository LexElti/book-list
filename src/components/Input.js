import React from 'react'
import {connect} from 'react-redux'
import {changeFilters} from '../store/actions'

export const Input = ({name, filters, changeFilters}) => {
  const changeHandler = event => {
    event.persist()
    const field = event.target.name 
    const value = event.target.value
    changeFilters(field, value)
  }

  return (
    <div className="col-lg-3  col-sm-6 mb-4">
      <input 
        type="text" 
        className="form-control" 
        name={name}
        value={filters[name]}
        placeholder={name.charAt(0).toUpperCase() + name.slice(1)}
        onChange={changeHandler}
      />
    </div>
  )
}

const mapStateToProps = state => ({
  filters: state.books.filters
})

const mapDispatchToProps = {
  changeFilters
}

export default connect(mapStateToProps, mapDispatchToProps)(Input)
