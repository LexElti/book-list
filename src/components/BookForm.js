import React from 'react'
import {connect} from 'react-redux' 
import {fetchBooks} from '../store/actions'
import Alert from './Alert'
import Input from './Input'

export const BookForm = ({filters, alert, fetchBooks}) => {
  const submitHandler = event => {
    event.preventDefault()
    fetchBooks(filters)
  }

  const inputs = Object.keys(filters)
    .map(key => <Input name={key} key={key} />)

  return (
    <form onSubmit={submitHandler}>
      {alert && <Alert text={alert} />}
      <div className="row">
        {inputs}
      </div>
      <button 
        className="btn btn-primary"
        type="submit"
      >Search</button>
    </form>
  )
}

const mapStateToProps = state => ({
  filters: state.books.filters,
  alert: state.app.alert
})

const mapDispatchToProps = {
  fetchBooks
}

export default connect(mapStateToProps, mapDispatchToProps)(BookForm)
