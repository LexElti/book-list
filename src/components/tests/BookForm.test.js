import React from 'react'
import {BookForm} from '../BookForm'
import Input from '../Input'
import Alert from '../Alert'

describe('<BookForm />', () => {
  const props = {
    filters: {
      title: '', author: '', publisher: '', subject: ''
    },
    alert: null,
    fetchBooks: () => {},
  }

  describe('<BookForm /> initial', () => {
    const wrapper = shallow(<BookForm {...props} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have components', () => {
      expect(wrapper.find(Input)).toHaveLength(4)
    })

    it('should not render Alert', () => {
      expect(wrapper.find(Alert)).toHaveLength(0)
    })
  })

  describe('<BookForm /> with warning message', () => {
    const nextProps = {
      ...props,
      alert: 'Something message',
    }

    const wrapper = shallow(<BookForm {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have components', () => {
      expect(wrapper.find(Input)).toHaveLength(4)
    })

    it('should render Alert', () => {
      expect(wrapper.find(Alert)).toHaveLength(1)
    })
  })

  describe('Form handlers', () => {
    const mockFetchBooks = jest.fn()
    
    const nextProps = {
      ...props,
      fetchBooks: mockFetchBooks
    }

    const wrapper = shallow(<BookForm {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('when submiting the form', () => {
      beforeEach(() => {
        wrapper.find('form').simulate('submit', {
          preventDefault: () => {}
        })
      })

      it('calls method fetchBooks', () => {
        expect(mockFetchBooks).toHaveBeenCalledTimes(1)
      })
    })

    describe('when clicking the button', () => {
      beforeEach(() => {
        wrapper.find('form').simulate('click', {
          preventDefault: () => {}
        })
      })

      it('calls method fetchBooks', () => {
        expect(mockFetchBooks).toHaveBeenCalledTimes(1)
      })
    })
  })
})
