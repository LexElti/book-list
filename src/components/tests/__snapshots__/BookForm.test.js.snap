// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookForm /> <BookForm /> initial renders properly 1`] = `ShallowWrapper {}`;

exports[`<BookForm /> <BookForm /> with warning message renders properly 1`] = `ShallowWrapper {}`;

exports[`<BookForm /> Form handlers renders properly 1`] = `ShallowWrapper {}`;
